### DISCUSSION TITLE
A few questions, if I may

### DISCUSSION THREAD

**Discussion by mogambro** [agnesoft/agdb#1061] [Q&A] [answered]
Hello!

Again: I love agdb, thanks so much for writing it! I dreamed of an embeddable graph db for many years, even had ideas to implement my own at some point, but I am no DB expert and have too many interests and not enough time these days, so thanks, really!

I am currently working on an application that uses AGDB, axum and leptos, so full-stack rust!

Now to the questions:

1) 

> is there a more appropriate place for questions? 
>
> Discussions are disabled on github, I saw icons for reddit and other social media, but no links?

2) 

> Should one store "blobs"?
> 
> In SQL databases sometimes blobs are stored (e.g. whole files) - does that make sense for a (file-based) AGDB?

3) 

> What's the maximum number of nodes/edges?
>
> i64 as DbId -> first bit decides whether edge or node -> 63 bits for information, 2^63 > 8 Quintillions
>
> 2^63=9,223,372,036,854,775,808
>
> So that number of edges and that number of nodes or is my math completely broken (again)?

4) 
> Is there a way to extract data easily for another backup, maybe in plain/text form or a way to store the decoded data > in sqlite etc.
>
> I am going all in with AGDB, but during development I encountered some problems in the past! Suddenly my graph could no longer be read. During development that's not a problem **and it didn't happen for a while now** (maybe it was because I forcefully stopped the application and there was a partial write - which should not be possible due to ACID guarantees...) in any case: agdb is not yet 1.0 and it's not backed by a huge team, so:

> Is there a way for my data to live safely and for me to sleep well? I love the queries and to work with AGDB, but once my application is done it has data in it that I cannot afford to lose.

**Comment by michaelvlach** [comment, marked as answer]
Hi, thank you for the support, it means a lot! As for your questions:

1. I have enabled discussions so feel free to use them. You can also open issues like this one, that is completely fine as well. The social media links are placeholders atm.

2. There is a binary data variant of `DbValue` which is internally `Vec<u8>` so blobs are supported and should be fine. The size should not be of too much concern unless of course the files are huge and many. While `agdb` can run without memory mapping it would be slower then and this does not sound like a good trade off. You would be probably better off storing a reference like a path to an image. On the other hand having everything in a single db file could be nice so if the data would not be overly large it should be ok. As a rule of thumb I ask myself if the value would be serachable. Obviously for binary files the answer is usually "no" but one can make it work. In SQL I always considered blob columns kind of anti-pattern but it really depends on the use case and size and number of them. Git also comes to mind thinking about it where storing large files is generally not advised but there is LFS so it can be done and sometimes is.

3. Your math is correct. I think you would sooner run out of storage (and definitely RAM if using memory mapping) before you would hit that limit. Also node ids are reused from the removed ones so the limit really is the full db of 2^63 nodes + 2^63 edges minus 1 because 0 is not used and indicates invalid index.

4. Recently I have added the `Db::copy()` and `Db::backup()` which should help here. At the moment there is no export functionality, I would need to come up with some format for that.

Regarding the data loss that is definitely important. WAL (the dot file alongside your main db file) should ensure that your db is always in a valid state and over the past versions I have fixed some issues with the db itself that could have cause data loss so maybe you were hit by one of them. Regardless there is I think one scenario where the current WAL could end up corrupting the db and that is if the WAL itself became corrupted and upon loading the db the broken WAL could break the main file's data. I have an idea how to prevent that though even if you were not hit by that specific scenario.

You are right that `agdb` is not yet 1.0 so I would very much recommend frequent backups & testing with your actual data. I do lots of testing as well but the real use case and specific data is always better. That's the only way how the 1.0 can be reached anyhow. :-)

**Comment by michaelvlach** [comment]
@nerdachse Hi again, I think I have fixed the potential issues around WAL. Details in the PR: https://github.com/agnesoft/agdb/pull/1062
